labs/b5.md

```diff
@@ -0,0 +1,364 @@
+
+Here the private network is on the left side of the router and the right side is the Internet. The router has an interface with a private IP address `192.168.1.1` and an interface with a public IP address `20.1.1.1`. 
+
+For IP packet 1, a host on the private network with address `192.168.1.3` tries to communicate with server B at public address `1.1.1.2` and creates a packet with the proper destination (dst) and source (src).  The packet arrives at the router and the router performs Source Network Address Translation (SNAT) changing the source IP on the packet to be `20.1.1.1` so that if Source B wants to respond it would send a packet to the router's public address since it cannot reach the host at its private IP address. This can be seen in IP packet 2. The packet arrive at the router again and the router now performs Destination Network Address Translation (DNAT) to change the source ip from the router's public address to the private IP address of the host that sent the packet.
+
+But how exactly does the router know which packets to convert to which address? A common procedure is for NAT routers to store a mapping of (`srcIP:srcPort`, `dstIP:dstPort`) so the router knows which machines are trying to communicate with one another and can translate packets that match these saved mappings.
+
+## DNS
+We've gone over IP addresses and how they are means of communicating with a host over IP, but  while IP addresses are machine friendly (computers love numbers) they aren't exactly human friendly. It's hard enough trying to remember phone numbers, memorizing 32 bit IP addresses isn't going to be any easier.
+
+But it's much easier for us to remember names like google, facebook, or coolmath-games.com. So out of this conflict the Domain Name System (DNS) was born as a compromise between machine friendly IP addresses and human friendly domain names.
+
+DNS is a system that maps a domain name like google.com to `172.217.6.78`. When you query for google.com your computer sends out a DNS query for google.com to a DNS server. Assuming things are properly configured and google.com has a valid corresponding address you will receive a response from an authoritative server that essentially says "google.com has IP address `x.x.x.x`".
+
+Now let's flush out this black magic a bit...
+
+## DNS Resolvers
+A DNS resolver on the client is responsible for performing DNS queries. A resolver refers to `/etc/resolv.conf` for nameserver information. The resolver then issues a query to the nameserver for the domain name. Additionally, `/etc/hosts.conf` stores static entries for resolving hostnames allowing a sysadmin to specify certain mapping manually. The `/etc/nsswitch.conf` file specifies the order in which a dns query performs name resolution, for example this line `hosts: files dns` in the file specifies that DNS lookups first try to find the name in the `/etc/hosts.conf` file before turning to nameservers listed in `/etc/resolv.conf`.
```

Comment:
just `/etc/hosts`

also mentioning nsswitch might be too much for basic.

Reply:
Sure I'll move this to the advanced version of the lab.